File: atlas-navigation/SLAM/FeatureExtractor.ts
/**
 * Feature Extractor - Extracts features from observations for SLAM
 */

import {
  Keypoint,
  Descriptor,
  FeatureExtractionParams,
  SLAMObservation,
} from "./SLAMTypes";

/**
 * Feature detector interface
 */
export interface Detector {
  detect(image: Uint8Array, width: number, height: number): Keypoint[];
}

/**
 * Descriptor extractor interface
 */
export interface DescriptorExtractor {
  compute(
    image: Uint8Array,
    width: number,
    height: number,
    keypoints: Keypoint[]
  ): number[][];
}

/**
 * Feature extraction result
 */
export interface FeatureExtractionResult {
  /** Detected keypoints */
  keypoints: Keypoint[];
  /** Computed descriptors */
  descriptors: number[][];
  /** Processing timestamp */
  timestamp: number;
}

/**
 * FeatureExtractor class for extracting features from observations
 */
export class FeatureExtractor {
  /** Feature detector instance */
  private detector: Detector;
  /** Descriptor extractor instance */
  private descriptorExtractor: DescriptorExtractor;
  /** Extraction parameters */
  private params: FeatureExtractionParams;
  /** Number of features extracted */
  private featureCount: number;
  /** Total extraction time accumulator */
  private totalExtractionTime: number;

  constructor(
    detector: Detector,
    descriptorExtractor: DescriptorExtractor,
    params?: Partial<FeatureExtractionParams>
  ) {
    this.detector = detector;
    this.descriptorExtractor = descriptorExtractor;
    this.params = {
      maxFeatures: params?.maxFeatures ?? 2000,
      pyramidLevels: params?.pyramidLevels ?? 4,
      scaleFactor: params?.scaleFactor ?? 1.2,
      fastThreshold: params?.fastThreshold ?? 20,
      minDistance: params?.minDistance ?? 5.0,
    };
    this.featureCount = 0;
    this.totalExtractionTime = 0;
  }

  /**
   * Extract features from an observation
   */
  extract(observation: SLAMObservation): FeatureExtractionResult {
    if (!observation.image) {
      throw new Error("Observation must contain image data for feature extraction");
    }

    const startTime = performance.now();

    // Detect keypoints
    const keypoints = this.detector.detect(
      observation.image.data,
      observation.image.width,
      observation.image.height
    );

    // Filter and refine keypoints
    const filteredKeypoints = this.filterKeypoints(keypoints);

    // Limit to max features
    const limitedKeypoints = filteredKeypoints.slice(0, this.params.maxFeatures);

    // Compute descriptors
    const descriptors = this.descriptorExtractor.compute(
      observation.image.data,
      observation.image.width,
      observation.image.height,
      limitedKeypoints
    );

    const endTime = performance.now();
    this.totalExtractionTime += endTime - startTime;
    this.featureCount += limitedKeypoints.length;

    return {
      keypoints: limitedKeypoints,
      descriptors,
      timestamp: observation.timestamp,
    };
  }

  /**
   * Detect keypoints in an image
   */
  detectKeypoints(image: Uint8Array, width: number, height: number): Keypoint[] {
    return this.detector.detect(image, width, height);
  }

  /**
   * Compute descriptors for given keypoints
   */
  computeDescriptors(
    image: Uint8Array,
    width: number,
    height: number,
    keypoints: Keypoint[]
  ): number[][] {
    return this.descriptorExtractor.compute(image, width, height, keypoints);
  }

  /**
   * Filter keypoints based on quality and minimum distance
   */
  private filterKeypoints(keypoints: Keypoint[]): Keypoint[] {
    // Sort by response strength (stronger features first)
    const sorted = [...keypoints].sort((a, b) => b.response - a.response);

    const result: Keypoint[] = [];

    for (const kp of sorted) {
      // Filter by minimum response
      if (kp.response < 1.0) {
        continue;
      }

      // Check minimum distance to existing keypoints
      let isFarEnough = true;
      for (const existing of result) {
        const dx = kp.pixel.x - existing.pixel.x;
        const dy = kp.pixel.y - existing.pixel.y;
        const distance = Math.sqrt(dx * dx + dy * dy);

        if (distance < this.params.minDistance) {
          isFarEnough = false;
          break;
        }
      }

      if (isFarEnough) {
        result.push(kp);
      }

      // Early exit if we have enough features
      if (result.length >= this.params.maxFeatures * 1.5) {
        break;
      }
    }

    return result;
  }

  /**
   * Get the detector instance
   */
  getDetector(): Detector {
    return this.detector;
  }

  /**
   * Get the descriptor extractor instance
   */
  getDescriptorExtractor(): DescriptorExtractor {
    return this.descriptorExtractor;
  }

  /**
   * Get feature extraction statistics
   */
  getStatistics(): {
    totalFeaturesExtracted: number;
    averageExtractionTime: number;
    params: FeatureExtractionParams;
  } {
    return {
      totalFeaturesExtracted: this.featureCount,
      averageExtractionTime:
        this.featureCount > 0 ? this.totalExtractionTime / this.featureCount : 0,
      params: this.params,
    };
  }

  /**
   * Reset statistics
   */
  resetStatistics(): void {
    this.featureCount = 0;
    this.totalExtractionTime = 0;
  }
}

/**
 * ORB detector implementation
 */
export class ORBDetector implements Detector {
  private fastThreshold: number;
  private pyramidLevels: number;
  private scaleFactor: number;

  constructor(
    fastThreshold: number = 20,
    pyramidLevels: number = 4,
    scaleFactor: number = 1.2
  ) {
    this.fastThreshold = fastThreshold;
    this.pyramidLevels = pyramidLevels;
    this.scaleFactor = scaleFactor;
  }

  detect(image: Uint8Array, width: number, height: number): Keypoint[] {
    // Simplified ORB detection using FAST corner detection
    // In a real implementation, this would use native libraries or WASM modules
    const keypoints: Keypoint[] = [];
    const bytesPerPixel = image.length / (width * height);

    for (let y = 8; y < height - 8; y += 4) {
      for (let x = 8; x < width - 8; x += 4) {
        // Simplified FAST-like corner response
        const cornerScore = this.computeCornerResponse(image, width, height, x, y, bytesPerPixel);

        if (cornerScore > this.fastThreshold) {
          // Estimate scale level from corner response
          const scaleLevel = this.estimateScaleLevel(cornerScore);

          // Compute orientation
          const angle = this.computeOrientation(image, width, height, x, y, bytesPerPixel);

          keypoints.push({
            pixel: { x, y },
            scaleLevel,
            response: cornerScore,
            angle,
            size: 7,
          });
        }
      }
    }

    // Limit keypoints if too many
    return keypoints.slice(0, 10000);
  }

  private computeCornerResponse(
    image: Uint8Array,
    width: number,
    height: number,
    cx: number,
    cy: number,
    bytesPerPixel: number
  ): number {
    // Simplified FAST-9 corner detector
    const centerIdx = (cy * width + cx) * bytesPerPixel;
    const intensity = image[centerIdx];

    let brighter = 0;
    let darker = 0;
    const circleOffsets = [
      [-3, 0],
      [-2, 2],
      [0, 3],
      [2, 2],
      [3, 0],
      [2, -2],
      [0, -3],
      [-2, -2],
    ];

    for (const [dx, dy] of circleOffsets) {
      const idx = ((cy + dy) * width + (cx + dx)) * bytesPerPixel;
      const diff = image[idx] - intensity;

      if (diff > 10) brighter++;
      else if (diff < -10) darker++;
    }

    // Return corner score based on consecutive pixels
    if (brighter >= 7) return 255 - intensity;
    if (darker >= 7) return intensity;

    return 0;
  }

  private estimateScaleLevel(cornerScore: number): number {
    // Map corner score to pyramid level
    return Math.floor(Math.log2(255 - cornerScore + 1)) % 4;
  }

  private computeOrientation(
    image: Uint8Array,
    width: number,
    height: number,
    cx: number,
    cy: number,
    bytesPerPixel: number
  ): number {
    // Compute first moments for orientation
    let m01 = 0;
    let m10 = 0;
    let m00 = 0;

    const patchSize = 7;

    for (let dy = -patchSize; dy <= patchSize; dy++) {
      for (let dx = -patchSize; dx <= patchSize; dx++) {
        const x = cx + dx;
        const y = cy + dy;

        if (x >= 0 && x < width && y >= 0 && y < height) {
          const idx = (y * width + x) * bytesPerPixel;
          const intensity = image[idx];

          m10 += intensity * dx;
          m01 += intensity * dy;
          m00 += intensity;
        }
      }
    }

    if (m00 === 0) return 0;

    // Compute angle
    return (Math.atan2(m01, m10) * 180) / Math.PI;
  }
}

/**
 * ORB descriptor extractor implementation
 */
export class ORBDescriptorExtractor implements DescriptorExtractor {
  private bytesPerDescriptor: number;

  constructor() {
    // ORB uses 256-bit (32 bytes) descriptors
    this.bytesPerDescriptor = 32;
  }

  compute(
    image: Uint8Array,
    width: number,
    height: number,
    keypoints: Keypoint[]
  ): number[][] {
    const descriptors: number[][] = [];

    for (const kp of keypoints) {
      const descriptor = this.computeDescriptor(image, width, height, kp);
      descriptors.push(descriptor);
    }

    return descriptors;
  }

  private computeDescriptor(
    image: Uint8Array,
    width: number,
    height: number,
    keypoint: Keypoint
  ): number[] {
    // Simplified ORB descriptor using BRIEF-like pattern
    const descriptor: number[] = [];
    const { x, y } = keypoint.pixel;
    const angleRad = (keypoint.angle * Math.PI) / 180;
    const cosA = Math.cos(angleRad);
    const sinA = Math.sin(angleRad);

    // BRIEF pattern pairs (simplified)
    const patternPairs = [
      [8, -7],
      [5, 5],
      [-2, 9],
      [-11, -2],
      [9, 4],
      [-6, -3],
      [7, -8],
      [3, -11],
      [0, 8],
      [-7, 3],
      [11, 0],
      [3, 7],
      [-8, -5],
      [6, 2],
      [-5, -9],
      [2, 10],
    ];

    for (const [dx1, dy1] of patternPairs) {
      // Rotate pattern by keypoint orientation
      const rx1 = Math.round(dx1 * cosA - dy1 * sinA);
      const ry1 = Math.round(dx1 * sinA + dy1 * cosA);

      const dx2 = dx1 + 1;
      const dy2 = dy1 + 1;
      const rx2 = Math.round(dx2 * cosA - dy2 * sinA);
      const ry2 = Math.round(dx2 * cosA + dy2 * sinA);

      // Sample intensities
      const x1 = Math.max(0, Math.min(width - 1, Math.round(x + rx1)));
      const y1 = Math.max(0, Math.min(height - 1, Math.round(y + ry1)));
      const x2 = Math.max(0, Math.min(width - 1, Math.round(x + rx2)));
      const y2 = Math.max(0, Math.min(height - 1, Math.round(y + ry2)));

      const idx1 = y1 * width + x1;
      const idx2 = y2 * width + x2;

      descriptor.push(image[idx1] < image[idx2] ? 1 : 0);
    }

    return descriptor;
  }
}

/**
 * Brute-force descriptor matcher
 */
export class BFMatcher {
  private normType: "HAMMING" | "L2";

  constructor(normType: "HAMMING" | "L2" = "HAMMING") {
    this.normType = normType;
  }

  /**
   * Match descriptors using brute-force
   */
  match(
    descriptors1: number[][],
    descriptors2: number[][]
  ): { queryIdx: number; trainIdx: number; distance: number }[] {
    const matches: { queryIdx: number; trainIdx: number; distance: number }[] = [];

    for (let i = 0; i < descriptors1.length; i++) {
      let minDist = Infinity;
      let minIdx = -1;

      for (let j = 0; j < descriptors2.length; j++) {
        const dist = this.computeDistance(descriptors1[i], descriptors2[j]);

        if (dist < minDist) {
          minDist = dist;
          minIdx = j;
        }
      }

      if (minIdx >= 0) {
        matches.push({ queryIdx: i, trainIdx: minIdx, distance: minDist });
      }
    }

    return matches;
  }

  /**
   * Match with ratio test (Lowe's ratio test)
   */
  matchWithRatio(
    descriptors1: number[][],
    descriptors2: number[][],
    ratioThreshold: number = 0.75
  ): { queryIdx: number; trainIdx: number; distance: number }[] {
    const matches: { queryIdx: number; trainIdx: number; distance: number }[] = [];

    for (let i = 0; i < descriptors1.length; i++) {
      // Find two nearest neighbors
      let bestDist = Infinity;
      let secondBestDist = Infinity;
      let bestIdx = -1;

      for (let j = 0; j < descriptors2.length; j++) {
        const dist = this.computeDistance(descriptors1[i], descriptors2[j]);

        if (dist < bestDist) {
          secondBestDist = bestDist;
          bestDist = dist;
          bestIdx = j;
        } else if (dist < secondBestDist) {
          secondBestDist = dist;
        }
      }

      // Apply ratio test
      if (bestIdx >= 0 && bestDist < secondBestDist * ratioThreshold) {
        matches.push({ queryIdx: i, trainIdx: bestIdx, distance: bestDist });
      }
    }

    return matches;
  }

  /**
   * Compute distance between two descriptors
   */
  private computeDistance(d1: number[], d2: number[]): number {
    if (this.normType === "HAMMING") {
      // Hamming distance for binary descriptors
      let dist = 0;
      for (let i = 0; i < d1.length && i < d2.length; i++) {
        if (d1[i] !== d2[i]) dist++;
      }
      return dist;
    } else {
      // Euclidean distance for float descriptors
      let dist = 0;
      for (let i = 0; i < d1.length && i < d2.length; i++) {
        const diff = d1[i] - d2[i];
        dist += diff * diff;
      }
      return Math.sqrt(dist);
    }
  }
}

File: atlas-navigation-cpp/src/FeatureExtractor.cpp
#include "atlas_navigation/SLAM/FeatureExtractor.h"
#include <algorithm>
#include <cmath>
#include <chrono>

namespace atlas_navigation {
namespace slam {

// --- ORBDetector ---

ORBDetector::ORBDetector(int fastThreshold, int pyramidLevels, double scaleFactor)
    : fastThreshold_(fastThreshold), pyramidLevels_(pyramidLevels), scaleFactor_(scaleFactor) {}

std::vector<Keypoint> ORBDetector::detect(const std::vector<uint8_t>& image, int width, int height) {
    std::vector<Keypoint> keypoints;
    int bytesPerPixel = static_cast<int>(image.size()) / (width * height);

    for (int y = 8; y < height - 8; y += 4) {
        for (int x = 8; x < width - 8; x += 4) {
            double cornerScore = computeCornerResponse(image, width, height, x, y, bytesPerPixel);
            if (cornerScore > fastThreshold_) {
                int scaleLevel = estimateScaleLevel(cornerScore);
                double angle = computeOrientation(image, width, height, x, y, bytesPerPixel);
                Keypoint kp;
                kp.pixel = {static_cast<double>(x), static_cast<double>(y)};
                kp.scaleLevel = scaleLevel;
                kp.response = cornerScore;
                kp.angle = angle;
                kp.size = 7.0;
                keypoints.push_back(kp);
            }
        }
    }

    if (keypoints.size() > 10000)
        keypoints.resize(10000);
    return keypoints;
}

double ORBDetector::computeCornerResponse(const std::vector<uint8_t>& image, int width, int height,
                                           int cx, int cy, int bytesPerPixel) const {
    int centerIdx = (cy * width + cx) * bytesPerPixel;
    uint8_t intensity = image[static_cast<size_t>(centerIdx)];
    int brighter = 0, darker = 0;
    const int circleOffsets[8][2] = {{-3,0},{-2,2},{0,3},{2,2},{3,0},{2,-2},{0,-3},{-2,-2}};

    for (auto& off : circleOffsets) {
        int idx = ((cy + off[1]) * width + (cx + off[0])) * bytesPerPixel;
        if (idx >= 0 && idx < static_cast<int>(image.size())) {
            int diff = static_cast<int>(image[static_cast<size_t>(idx)]) - intensity;
            if (diff > 10) brighter++;
            else if (diff < -10) darker++;
        }
    }

    if (brighter >= 7) return 255.0 - intensity;
    if (darker >= 7) return static_cast<double>(intensity);
    return 0.0;
}

int ORBDetector::estimateScaleLevel(double cornerScore) const {
    return static_cast<int>(std::log2(255.0 - cornerScore + 1.0)) % 4;
}

double ORBDetector::computeOrientation(const std::vector<uint8_t>& image, int width, int height,
                                        int cx, int cy, int bytesPerPixel) const {
    double m01 = 0, m10 = 0, m00 = 0;
    int patchSize = 7;

    for (int dy = -patchSize; dy <= patchSize; ++dy) {
        for (int dx = -patchSize; dx <= patchSize; ++dx) {
            int x = cx + dx, y = cy + dy;
            if (x >= 0 && x < width && y >= 0 && y < height) {
                size_t idx = static_cast<size_t>((y * width + x) * bytesPerPixel);
                double intensity = image[idx];
                m10 += intensity * dx;
                m01 += intensity * dy;
                m00 += intensity;
            }
        }
    }

    if (m00 == 0) return 0;
    return std::atan2(m01, m10) * 180.0 / M_PI;
}

// --- ORBDescriptorExtractor ---

ORBDescriptorExtractor::ORBDescriptorExtractor() = default;

std::vector<std::vector<double>> ORBDescriptorExtractor::compute(
    const std::vector<uint8_t>& image, int width, int height,
    const std::vector<Keypoint>& keypoints) {
    std::vector<std::vector<double>> descriptors;
    descriptors.reserve(keypoints.size());
    for (const auto& kp : keypoints) {
        descriptors.push_back(computeDescriptor(image, width, height, kp));
    }
    return descriptors;
}

std::vector<double> ORBDescriptorExtractor::computeDescriptor(
    const std::vector<uint8_t>& image, int width, int height,
    const Keypoint& keypoint) const {
    std::vector<double> descriptor;
    double x = keypoint.pixel.first, y = keypoint.pixel.second;
    double angleRad = keypoint.angle * M_PI / 180.0;
    double cosA = std::cos(angleRad), sinA = std::sin(angleRad);

    const int patternPairs[16][2] = {
        {8,-7},{5,5},{-2,9},{-11,-2},{9,4},{-6,-3},{7,-8},{3,-11},
        {0,8},{-7,3},{11,0},{3,7},{-8,-5},{6,2},{-5,-9},{2,10}
    };

    for (auto& pp : patternPairs) {
        int rx1 = static_cast<int>(std::round(pp[0] * cosA - pp[1] * sinA));
        int ry1 = static_cast<int>(std::round(pp[0] * sinA + pp[1] * cosA));
        int rx2 = static_cast<int>(std::round((pp[0]+1) * cosA - (pp[1]+1) * sinA));
        int ry2 = static_cast<int>(std::round((pp[0]+1) * sinA + (pp[1]+1) * cosA));

        int x1 = std::max(0, std::min(width - 1, static_cast<int>(std::round(x + rx1))));
        int y1 = std::max(0, std::min(height - 1, static_cast<int>(std::round(y + ry1))));
        int x2 = std::max(0, std::min(width - 1, static_cast<int>(std::round(x + rx2))));
        int y2 = std::max(0, std::min(height - 1, static_cast<int>(std::round(y + ry2))));

        uint8_t v1 = image[static_cast<size_t>(y1 * width + x1)];
        uint8_t v2 = image[static_cast<size_t>(y2 * width + x2)];
        descriptor.push_back(v1 < v2 ? 1.0 : 0.0);
    }
    return descriptor;
}

// --- BFMatcher ---

BFMatcher::BFMatcher(NormType normType) : normType_(normType) {}

std::vector<BFMatcher::Match> BFMatcher::match(
    const std::vector<std::vector<double>>& descriptors1,
    const std::vector<std::vector<double>>& descriptors2) const {
    std::vector<Match> matches;
    for (size_t i = 0; i < descriptors1.size(); ++i) {
        double minDist = std::numeric_limits<double>::max();
        int minIdx = -1;
        for (size_t j = 0; j < descriptors2.size(); ++j) {
            double dist = computeDistance(descriptors1[i], descriptors2[j]);
            if (dist < minDist) { minDist = dist; minIdx = static_cast<int>(j); }
        }
        if (minIdx >= 0) matches.push_back({static_cast<int>(i), minIdx, minDist});
    }
    return matches;
}

std::vector<BFMatcher::Match> BFMatcher::matchWithRatio(
    const std::vector<std::vector<double>>& descriptors1,
    const std::vector<std::vector<double>>& descriptors2,
    double ratioThreshold) const {
    std::vector<Match> matches;
    for (size_t i = 0; i < descriptors1.size(); ++i) {
        double bestDist = std::numeric_limits<double>::max();
        double secondBestDist = std::numeric_limits<double>::max();
        int bestIdx = -1;
        for (size_t j = 0; j < descriptors2.size(); ++j) {
            double dist = computeDistance(descriptors1[i], descriptors2[j]);
            if (dist < bestDist) { secondBestDist = bestDist; bestDist = dist; bestIdx = static_cast<int>(j); }
            else if (dist < secondBestDist) { secondBestDist = dist; }
        }
        if (bestIdx >= 0 && bestDist < secondBestDist * ratioThreshold) {
            matches.push_back({static_cast<int>(i), bestIdx, bestDist});
        }
    }
    return matches;
}

double BFMatcher::computeDistance(const std::vector<double>& d1, const std::vector<double>& d2) const {
    if (normType_ == HAMMING) {
        double dist = 0;
        size_t n = std::min(d1.size(), d2.size());
        for (size_t i = 0; i < n; ++i) {
            if (d1[i] != d2[i]) dist++;
        }
        return dist;
    } else {
        double dist = 0;
        size_t n = std::min(d1.size(), d2.size());
        for (size_t i = 0; i < n; ++i) {
            double diff = d1[i] - d2[i];
            dist += diff * diff;
        }
        return std::sqrt(dist);
    }
}

// --- FeatureExtractor ---

FeatureExtractor::FeatureExtractor(std::unique_ptr<Detector> detector,
                                     std::unique_ptr<DescriptorExtractor> descriptorExtractor,
                                     const FeatureExtractionParams& params)
    : detector_(std::move(detector)), descriptorExtractor_(std::move(descriptorExtractor)), params_(params) {}

FeatureExtractionResult FeatureExtractor::extract(const SLAMObservation& observation) {
    if (!observation.image.has_value()) {
        // Throw logic_error instead
        throw std::logic_error("Observation must contain image data for feature extraction");
    }

    auto startTime = std::chrono::steady_clock::now();
    const auto& img = observation.image.value();

    auto keypoints = detector_->detect(img.data, img.width, img.height);
    auto filtered = filterKeypoints(keypoints);

    if (static_cast<int>(filtered.size()) > params_.maxFeatures)
        filtered.resize(static_cast<size_t>(params_.maxFeatures));

    auto descriptors = descriptorExtractor_->compute(img.data, img.width, img.height, filtered);

    auto endTime = std::chrono::steady_clock::now();
    double elapsed = std::chrono::duration<double, std::milli>(endTime - startTime).count();
    totalExtractionTime_ += elapsed;
    featureCount_ += static_cast<int>(filtered.size());

    FeatureExtractionResult result;
    result.keypoints = std::move(filtered);
    result.descriptors = std::move(descriptors);
    result.timestamp = observation.timestamp;
    return result;
}

std::vector<Keypoint> FeatureExtractor::detectKeypoints(const std::vector<uint8_t>& image, int width, int height) {
    return detector_->detect(image, width, height);
}

std::vector<std::vector<double>> FeatureExtractor::computeDescriptors(
    const std::vector<uint8_t>& image, int width, int height,
    const std::vector<Keypoint>& keypoints) {
    return descriptorExtractor_->compute(image, width, height, keypoints);
}

std::vector<Keypoint> FeatureExtractor::filterKeypoints(const std::vector<Keypoint>& keypoints) const {
    auto sorted = keypoints;
    std::sort(sorted.begin(), sorted.end(), [](const Keypoint& a, const Keypoint& b) {
        return a.response > b.response;
    });

    std::vector<Keypoint> result;
    for (const auto& kp : sorted) {
        if (kp.response < 1.0) continue;
        bool isFarEnough = true;
        for (const auto& existing : result) {
            double dx = kp.pixel.first - existing.pixel.first;
            double dy = kp.pixel.second - existing.pixel.second;
            if (std::sqrt(dx*dx + dy*dy) < params_.minDistance) {
                isFarEnough = false;
                break;
            }
        }
        if (isFarEnough) {
            result.push_back(kp);
        }
        if (static_cast<int>(result.size()) >= static_cast<int>(params_.maxFeatures * 1.5))
            break;
    }
    return result;
}

FeatureExtractor::Statistics FeatureExtractor::getStatistics() const {
    Statistics s;
    s.totalFeaturesExtracted = featureCount_;
    s.averageExtractionTime = featureCount_ > 0 ? totalExtractionTime_ / featureCount_ : 0.0;
    s.params = params_;
    return s;
}

void FeatureExtractor::resetStatistics() {
    featureCount_ = 0;
    totalExtractionTime_ = 0.0;
}

} // namespace slam
} // namespace atlas_navigation
